### PR TITLE
feat: Data.Fin.Lemmas <- mathlib

### PR DESCRIPTION
**Pull request by digama0**
A WIP branch for porting Mathlib.Data.Fin.Basic to std.

### CODE REVIEW

**Comment by fgdorais** [comment]
What kind of assistance do you require?

**Comment by digama0** [comment]
It's basically unfinished work, I made it about 1/4 of the way through the file. The task here is to remove anything referencing mathlib notions like `NeZero` or `Equiv` and otherwise de-mathlibify all the other lemmas in the file. I haven't looked too carefully at the tail end of the file, maybe there is something that shouldn't be translated, but I would expect roughly the whole thing can move to std. This is relatively high priority since the mathlib std4 bump is blocked on this (in part).

File: Std/Data/Fin/Lemmas.lean [hunk 1]
```diff
@@ -28,6 +66,669 @@ theorem lt_def {a b : Fin n} : a < b ↔ a.1 < b.1 := .rfl
...
+    natAdd n (subNat n (cast (Nat.add_comm ..) i) h) = i := by simp [← cast_addNat]; rfl
+
+/-! ### recursion and induction principles -/
+
+/-- Define `C n i` by induction on `i : Fin n` interpreted as `(0 : Fin (n - i)).succ.succ…`.
+This function has two arguments: `H0 n` defines `0`-th element `C (n+1) 0` of an `(n+1)`-tuple,
+and `Hs n i` defines `(i+1)`-st element of `(n+1)`-tuple based on `n`, `i`, and `i`-th element
+of `n`-tuple. -/
+-- FIXME: Performance review
+@[elab_as_elim] def succRec {C : ∀ n, Fin n → Sort _} (H0 : ∀ n, C n.succ (0 : Fin (n + 1)))
```
Comment (fgdorais) [inline Std/Data/Fin/Lemmas.lean:555]: Hypotheses: `H0` -> `zero`, `Hs` -> `succ`

File: Std/Data/Fin/Lemmas.lean [hunk 2]
```diff
@@ -28,6 +66,669 @@ theorem lt_def {a b : Fin n} : a < b ↔ a.1 < b.1 := .rfl
...
+  | Nat.succ _, ⟨Nat.succ i, h⟩ => Hs _ _ (succRec H0 Hs ⟨i, Nat.lt_of_succ_lt_succ h⟩)
+
+/-- Define `C n i` by induction on `i : Fin n` interpreted as `(0 : Fin (n - i)).succ.succ…`.
+This function has two arguments: `H0 n` defines `0`-th element `C (n+1) 0` of an `(n+1)`-tuple,
+and `Hs n i` defines `(i+1)`-st element of `(n+1)`-tuple based on `n`, `i`, and `i`-th element
+of `n`-tuple.
+
+A version of `Fin.succRec` taking `i : Fin n` as the first argument. -/
+-- FIXME: Performance review
+@[elab_as_elim] def succRecOn {n : Nat} (i : Fin n) {C : ∀ n, Fin n → Sort _}
```
Comment (fgdorais) [inline Std/Data/Fin/Lemmas.lean:568]: Hypotheses: `H0` -> `zero`, `Hs` -> `succ`

File: Std/Data/Fin/Lemmas.lean [hunk 3]
```diff
@@ -28,6 +66,669 @@ theorem lt_def {a b : Fin n} : a < b ↔ a.1 < b.1 := .rfl
...
+
+@[simp] theorem succRecOn_succ {C : ∀ n, Fin n → Sort _} {H0 Hs} {n} (i : Fin n) :
+    @Fin.succRecOn (n + 1) i.succ C H0 Hs = Hs n i (Fin.succRecOn i H0 Hs) := by cases i; rfl
+
+/-- Define `C i` by induction on `i : Fin (n + 1)` via induction on the underlying `Nat` value.
+This function has two arguments: `h0` handles the base case on `C 0`,
+and `hs` defines the inductive step using `C i.castSucc`.
+-/
+-- FIXME: Performance review
+@[elab_as_elim] def induction {C : Fin (n + 1) → Sort _} (h0 : C 0)
```
Comment (fgdorais) [inline Std/Data/Fin/Lemmas.lean:583]: Hypotheses: `h0` -> `zero`, `hs` -> `succ`

File: Std/Data/Fin/Lemmas.lean [hunk 4]
```diff
@@ -28,6 +66,669 @@ theorem lt_def {a b : Fin n} : a < b ↔ a.1 < b.1 := .rfl
...
+    (induction h0 hs : ∀ i : Fin (n+1), C i) i.succ = hs i (induction h0 hs (castSucc i)) := rfl
+
+/-- Define `C i` by induction on `i : Fin (n + 1)` via induction on the underlying `Nat` value.
+This function has two arguments: `h0` handles the base case on `C 0`,
+and `hs` defines the inductive step using `C i.castSucc`.
+
+A version of `Fin.induction` taking `i : Fin (n + 1)` as the first argument.
+-/
+-- FIXME: Performance review
+@[elab_as_elim] def inductionOn (i : Fin (n + 1)) {C : Fin (n + 1) → Sort _} (h0 : C 0)
```
Comment (fgdorais) [inline Std/Data/Fin/Lemmas.lean:604]: Hypotheses: `h0` -> `zero`, `hs` -> `succ`

File: Std/Data/Fin/Lemmas.lean [hunk 5]
```diff
@@ -28,6 +66,669 @@ theorem lt_def {a b : Fin n} : a < b ↔ a.1 < b.1 := .rfl
...
+    (hs : ∀ i : Fin n, C (castSucc i) → C i.succ) : C i := induction h0 hs i
+
+/-- Define `f : Π i : Fin n.succ, C i` by separately handling the cases `i = 0` and
+`i = j.succ`, `j : Fin n`. -/
+@[elab_as_elim] def cases {C : Fin (n + 1) → Sort _} (H0 : C 0) (Hs : ∀ i : Fin n, C i.succ) :
```
Comment (fgdorais) [inline Std/Data/Fin/Lemmas.lean:609]: Hypotheses: `H0` -> `zero`, `Hs` -> `succ`

File: Std/Data/Fin/Lemmas.lean [hunk 6]
```diff
@@ -28,6 +66,669 @@ theorem lt_def {a b : Fin n} : a < b ↔ a.1 < b.1 := .rfl
...
+
+theorem fin_two_eq_of_eq_zero_iff : ∀ {a b : Fin 2}, (a = 0 ↔ b = 0) → a = b := by
+  simp [forall_fin_two]
+
+/--
+Define `C i` by reverse induction on `i : Fin (n + 1)` via induction on the underlying `Nat` value.
+This function has two arguments: `hlast` handles the base case on `C (Fin.last n)`,
+and `hs` defines the inductive step using `C i.succ`, inducting downwards.
+-/
+@[elab_as_elim] def reverseInduction {C : Fin (n + 1) → Sort _} (hlast : C (Fin.last n))
```
Comment (fgdorais) [inline Std/Data/Fin/Lemmas.lean:649]: Hypotheses: `hlast` -> `last`, `hs` -> `cast`

File: Std/Data/Fin/Lemmas.lean [hunk 7]
```diff
@@ -28,6 +66,669 @@ theorem lt_def {a b : Fin n} : a < b ↔ a.1 < b.1 := .rfl
...
+  rw [reverseInduction]; simp; rfl
+
+@[simp] theorem reverse_induction_castSucc {n : Nat} {C : Fin (n + 1) → Sort _}
+    (h0 : C (Fin.last n)) (hs : ∀ i : Fin n, C i.succ → C (castSucc i)) (i : Fin n) :
+    reverseInduction h0 hs (castSucc i) = hs i (reverseInduction h0 hs i.succ) := by
+  rw [reverseInduction, dif_neg (Fin.ne_of_lt (Fin.castSucc_lt_last i))]; rfl
+
+/-- Define `f : Π i : Fin n.succ, C i` by separately handling the cases `i = Fin.last n` and
+`i = j.castSucc`, `j : Fin n`. -/
+@[elab_as_elim] def lastCases {n : Nat} {C : Fin (n + 1) → Sort _} (hlast : C (Fin.last n))
```
Comment (fgdorais) [inline Std/Data/Fin/Lemmas.lean:672]: Hypotheses: `hlast` -> `last`, `hcast` -> `cast`

File: Std/Data/Fin/Lemmas.lean [hunk 8]
```diff
@@ -28,6 +66,669 @@ theorem lt_def {a b : Fin n} : a < b ↔ a.1 < b.1 := .rfl
...
+  reverse_induction_last _ _
+
+@[simp] theorem lastCases_castSucc {n : Nat} {C : Fin (n + 1) → Sort _} (hlast : C (Fin.last n))
+    (hcast : ∀ i : Fin n, C (castSucc i)) (i : Fin n) :
+    (Fin.lastCases hlast hcast (Fin.castSucc i) : C (Fin.castSucc i)) = hcast i :=
+  reverse_induction_castSucc _ _ _
+
+/-- Define `f : Π i : Fin (m + n), C i` by separately handling the cases `i = castAdd n i`,
+`j : Fin m` and `i = natAdd m j`, `j : Fin n`. -/
+@[elab_as_elim] def addCases {m n : Nat} {C : Fin (m + n) → Sort u} (hleft : ∀ i, C (castAdd n i))
```
Comment (fgdorais) [inline Std/Data/Fin/Lemmas.lean:688]: Hypotheses: `hleft` -> `left`, `hright` -> `right`

**Review by fgdorais — CHANGES_REQUESTED** [review]
Custom recursors work better when the hypotheses have predictable and meaningful names. This is to facilitate the use of the pattern `induction x using foo with | hyp ... => ...` I went through and recommended better names.

PS: Also consider using `motive` instead of `C` but I don't think that's essential.

**Comment by kmill** [comment]
The de-ord-embedding/iso-ification seems like it was done right, so here's a :+1: from me, though I only checked about 400-500 lines in detail

**Review by fgdorais — APPROVED** [review]

**Pull request merged**